### PR TITLE
NIP-42 Auth for REQs

### PR DESCRIPTION
**Pull request by mplorentz**
## Issues covered
#1325 

## Description
This handles `AUTH` messages from relays according to [NIP-42](https://github.com/nostr-protocol/nips/blob/master/42.md). When we receive an `AUTH` message from a relay we put it into an `authenticating` state, close any open subscriptions, and send the signed challenge. The `authenticating` state will prevent new `REQ`s from being sent to the relay until we receive an `OK` message that our challenge event was accepted. The socket is then moved to the `.connected` state and all subscriptions are reopened.

This works well for reading events from the relay, but does not handle writes very gracefully. Currently we queue events for publishing by writing them to disk and then retrying them every 60 seconds until they publish successfully. This means that if the socket isn't already authenticated then the publish will fail and we'll retry again later. All the relays I have tested send `AUTH` immediately when you connect, which means the writes should succeed unless you have just launched the app or come back online. But I would like to write a more robust system that queues the writes and waits until we we are connected and authenticated to send them. However I'd like to do that in a separate PR since it's a bunch of additional work and it's not required for #1311.

## How to test
I had trouble finding a relay to test this code with. All relays that I know of only require AUTH for writes, and you have to pay to write to them. `wss://relay.nostr.wine` is expensive and I bought more sats to join it but they aren't spendable yet 🙄. `wss://eden.nostr.land` will do the AUTH dance with Nos successfully but it will let you read even without it, and I tried paying some sats so that I could write to it but it didn't work 😭 the site appears to be broken.

The best way to test is to make sure you can read events from `wss://olympics2024.nos.social`. Event `note1rkx7mcg8k28dr53raaap5ylxls9emuwxa36ntdgu3trfrhm8zc8sk9rg9q` published by me with content "This event should only be visible on olympics2024.nos.social" is a good one to test with. Note that publishing to a single relay from Nos is currently broken but fixed in #1333.

### CODE REVIEW

**Comment by mplorentz** [comment]
@joshuatbrown I'm marking this ready for review even though the olympics relay isn't ready for testing yet. I'd like to get feedback on the code and as much of the functionality as you are able to test. Hoping that once the relay is up we can do some quick testing with it and get this merged.

**Comment by mplorentz** [comment]
Ok the olympics relay is up! Ready for review.

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+- Added support for paid and authenticated relays (NIP-42).
```
Comment (joshuatbrown) [inline CHANGELOG.md:11]: Do want to add NIP-42 to our README as well?

File: Nos/Service/Relay/RelayService.swift [hunk 2]
```diff
@@ -473,6 +483,52 @@ extension RelayService {
...
     }
     
+    private func handleClosed(from socket: WebSocket, responseArray: [Any]) async {
+        guard responseArray.count > 1 else {
+            return
+        }
+        
+        if let subID = responseArray[1] as? RelaySubscription.ID {
+            await subscriptionManager.receivedClose(for: subID, from: socket)
+        }
```
Comment (joshuatbrown) [inline Nos/Service/Relay/RelayService.swift:493]: Could this all be simplified like the code below? Something like:

```
guard let subID = responseArray[safe: 1] as? RelaySubscription.ID else {
    return
}
await subscriptionManager...
```

File: Nos/Service/Relay/RelaySubscriptionManager.swift [hunk 3]
```diff
@@ -179,6 +182,40 @@ actor RelaySubscriptionManagerActor: RelaySubscriptionManager {
...
+                    socket.write(string: "[\"CLOSE\", \"\(subscription.id)\"]")
+                }
+            }
+            connection.state = .authenticating(responseID)
+        }
+    }
+    
+    /// Checks the ID of an "OK" message from a relay to see if it matches any authentication events we have sent.
+    /// If it does we mark the relay as .connected. If it doesn't then we do nothing, so this function is safe to be 
+    /// called on every "OK" message.
```
Comment (joshuatbrown) [inline Nos/Service/Relay/RelaySubscriptionManager.swift:201]: Excellent docs.

File: Nos/Service/Relay/RelaySubscriptionManager.swift [hunk 4]
```diff
@@ -179,6 +182,40 @@ actor RelaySubscriptionManagerActor: RelaySubscriptionManager {
...
+            }
+            connection.state = .authenticating(responseID)
+        }
+    }
+    
+    /// Checks the ID of an "OK" message from a relay to see if it matches any authentication events we have sent.
+    /// If it does we mark the relay as .connected. If it doesn't then we do nothing, so this function is safe to be 
+    /// called on every "OK" message. 
+    func checkAuthentication(success: Bool, from socket: WebSocket, eventID: RawNostrID, message: String?) -> Bool {
+        if let relayAddress = socket.url, let connection = socketConnections[relayAddress] {
```
Comment (joshuatbrown) [inline Nos/Service/Relay/RelaySubscriptionManager.swift:203]: Could we flip this statement to a `guard` to avoid the three levels of `if` statements?

**Comment by mplorentz** [comment]
> Oh this is interesting. I added the olympics2024.nos.social relay to my list with my "I am Groot" account (npub1qfn650vj62j8ntttenwxlem9wqmaa26tw75thn7nvcasamceddvqy2zdu4). I'm able to see your event in my feed.

@joshuatbrown This is a known issue with the relay that we are fixing now. Only keys in the allowlist should be able to read the event, and I am Groot is not afaik.

> the search box fails when I paste in note1rkx7mcg8k28dr53raaap5ylxls9emuwxa36ntdgu3trfrhm8zc8sk9rg9q

Looks like this is an issue on main.

> So I can see note1rkx7mcg8k28dr53raaap5ylxls9emuwxa36ntdgu3trfrhm8zc8sk9rg9q with my personal account (npub1yl8jc6znttslcpj3p6p8vuq98awu6w0xh4lqtu0lkjr772kpx4ysfqvz34) even when I don't have wss://olympics2024.nos.social in my relay list

is the event cached in Core Data? Like did you previously connect to the olympics relay on that installation? If this note has already leaked to other relays somehow I'm going to be mad 😅

File: CHANGELOG.md (hunk 1 above)
Comment (mplorentz) [reply CHANGELOG.md:11]: 💯

File: Nos/Service/Relay/RelayService.swift [hunk 5]
```diff
@@ -407,23 +404,36 @@ extension RelayService {
...
-            let eventId = responseArray[1] as? String,
-            let socketUrl = socket.request.url?.absoluteString {
+            let eventID = responseArray[1] as? String,
+            let socketURL = socket.request.url?.absoluteString {
+            
+            if await subscriptionManager.checkAuthentication(
+                success: success, 
+                from: socket, 
+                eventID: eventID, 
+                message: responseArray[3] as? String
```
[the remark this answers is not in the document]
Comment (mplorentz) [reply Nos/Service/Relay/RelayService.swift:425]: I'll try to make it clearer.

**Review by joshuatbrown — APPROVED** [review]
🥇 Code all looks good, and I'm able to see the magical event on the relay.

**Comment by joshuatbrown** [comment]
> > So I can see note1rkx7mcg8k28dr53raaap5ylxls9emuwxa36ntdgu3trfrhm8zc8sk9rg9q with my personal account (npub1yl8jc6znttslcpj3p6p8vuq98awu6w0xh4lqtu0lkjr772kpx4ysfqvz34) even when I don't have wss://olympics2024.nos.social in my relay list
> 
> is the event cached in Core Data? Like did you previously connect to the olympics relay on that installation? If this note has already leaked to other relays somehow I'm going to be mad 😅

That's probably what happened. I thought I had erase = true in PersistenceController but who knows?

**Pull request merged**